src/osd/OSDMap.cc

```diff
@@ -3261,6 +3261,27 @@ bool OSDMap::crush_rule_in_use(int rule_id) const
   return false;
 }
 
+int OSDMap::validate_crush_rules(CrushWrapper *newcrush,
+				 ostream *ss) const
+{
+  for (auto& i : pools) {
+    auto& pool = i.second;
+    int ruleno = pool.get_crush_rule();
+    if (!newcrush->rule_exists(ruleno)) {
+      if (ss)
+	*ss << "pool " << i.first << " references crush_rule " << ruleno
+	    << " but it is not present";
+      return -EINVAL;
+    }
+    if (newcrush->get_rule_mask_ruleset(ruleno) != ruleno) {
+      if (ss)
+	*ss << "rule " << ruleno << " mask ruleset does not match rule id";
+      return -EINVAL;
```

Comment:
can we also check
```c++

if (!newcrush->check_crush_rule(ruleno, pool.get_type(), pool.get_size(), *ss)) {
  return -EINVAL;
}
```
? so we can verify the type and size range of the rule are sane.

but we might need to update `check_crush_rule()`, so it prints out the ruleid in the error message.

Reply:
updated = validate to make sure min/max size are ok